### ISSUE TITLE
don't agree with 3.9 ("Require modules by folders, as opposed to the files directly")

### ISSUE THREAD

**Issue by jimmywarting** [goldbergyoni/nodebestpractices#824]
you will have a footgun for using `require('./index')` or just `require('./')` and nameless extension if you ever decide to switch to es7 modules with import or changing platform to browser module or Deno land. `require` don't work in browser or Deno. if you are building cross different environments, then you better use `import 'fullpath.js'` for an easier transition and better compatibility 

Ryan dahl regrets every bit of nodes module resolver...

**Comment by elektronik2k5** [comment]
I agree and ban `index.js/ts/whatever` files in the projects which I lead for this and other reasons.

**Comment by goldbergyoni** [comment]
Hey @jimmywarting, welcome aboard! 🌻

Well, one thing for sure - When I wrote this 4 years ago I felt much more confident with this advice. Things got more complicated now with the addition of ESM and I'm less sure about it. At the same time, I'm not sure about the opposite way around.

I'm putting Deno (a premature platform that might have a bright future or not) and Ryan Dahl (who regrets many other things and might regret many others in the future) out of this discussion.

There is a strong reason to expose module internals through index - Encapsulation. On the other end, browser compatibility is a thing for some teams, while not relevant for others. Couple of thoughts from the top of my head:

1. Putting aside the small implementation details (nameless extension, specifier-resolution), is this a good pattern - A team might decide to use index with explicit paths (/package/index.js), so is this a good pattern to use? I suggest that absolutely YES. It acts as an interface that spares from the client to dig deep inside for relevant files. Any objections to this? On the other end, calling a mediating file from the frontend might lead to performance issues which can be solved with some webpack love

2. Implementation details - Should a team needs to maintain browser-compatibility then implicit index resolution should be avoided indeed. But what if it doesn't (and fancy the value of bullet 1)? Should a team who already uses 'require' do it in a ESM-compatible way given that ESM is still youngsupportand many are still waiting? Why? In any case, when migrating to ESM in the future, they would need to rewrite the 'require' code (unless I miss anything). One more consideration for a backend-only team is that currently Node supports specifier-resolution flag that allows specifying index as the default for ESM

Those are just my raw thoughts that are brought to enrich the decision and help us make updated advice in this regard. Thanks for bringing this interesting topic.

cc @giltayar who might find this topic interesting

**Comment by giltayar** [comment]
Just resaw this. I most definitely think it's time to remove this section, given that ES modules _do not allow this behavior_. In ESM, if you `import(./folder)` it will NOT search for `./folder/index.js` (or `./folder/package.json` and then `main`). In ESM you MUST specify the exact relative path, including the extension.

**Comment by goldbergyoni** [comment]
@giltayar No doubt that white ESM walkers are coming to winterfell and these import files won't work. That said, regardless of the mechanics, I still think that the text suggests a valuable thing 👇 (copied from bullet 3.9):

**TL;DR:** When developing a module/library in a folder, place a facade file (e.g. index.js) that exposes the module's internals so every consumer will pass through it. This serves as an 'interface' to your module and eases future changes without breaking the contract

Otherwise: Changing the internal structure of files or the signature may break the interface with clients

3.9 Code example

```
// Do
const {SMSSender,  SMSNumberValidator} = require("./SMSProvider/index.js");

// Avoid
const SMSSender = require("./SMSProvider/SMSProvider.js");
const SMSValidator = require("./SMSProvider/validation/international/sms-validator.js");

```

Thoughts?

p.s. Your name sounds familiar, did you by chance speak at Node.JS TLV 2020?

**Comment by elektronik2k5** [comment]
I suggest one seemingly small and yet important enough (IMHO) change:
```diff
 // Do
const {SMSSender,  SMSNumberValidator} = require("./SMSProvider/index.js");
+ // Or
+ const SMSSender = require("./SMSProvider/SMSProvider.js");

// Avoid
- const SMSSender = require("./SMSProvider/SMSProvider.js");
const SMSValidator = require("./SMSProvider/validation/international/sms-validator.js");
```

There are actually two distinct, yet related points here:
1. I think that `index.{js,ts,css,etc}` files are harmful and ban them in projects which I lead. I realize that's highly opinionated and tons of people and projects would disagree. That's why in the BP I'd like to offer an alternative:
2. Correct me if I'm wrong, but the whole point of this BP is separation of public from private API's - as opposed to encouraging any specific naming schema. As such, the naming of this public file isn't important and we should offer a few examples in addition to the standard `index.js`.

**Comment by jimmywarting** [comment]
i don't think it's nothing wrong in using the filename `index.js` everything that is allowed to do with ESM paths (without compiler such as babel/ts/or node module resolver interfere) should be allowed. think of it how it would work in browser with type=module
You would have to write `require('./index.js)` with extension then...

I'm no fan of enforcing style rules or other best practices (this usually makes me frustrated). IMO i think a linter should only spot errors.

**Comment by giltayar** [comment]
> When developing a module/library in a folder, place a facade file (e.g. index.js) that exposes the module's internals so every consumer will pass through it. This serves as an 'interface' to your module and eases future changes without breaking the contract

This just won't work in ESM. It can't. But! There's an alternative option, which will work for CJS _and_ ESM, and is now available in all supported version of Node.js (ie 12 and up): `exports`. https://nodejs.org/api/packages.html#packages_package_entry_points

I would suggest to change the text and examples to use `exports`.

**Comment by jimmywarting** [comment]
After jumping on the Deno wagon i started to change my mind about `index.js` - it's a bad name
https://deno.land/manual/contributing/style_guide#do-not-use-the-filename-codeindextscodecodeindexjscode

while you may do the correct thing about importing files in the correct way with extension and how ESM works properly, you will start to see other abusing your index file in the wrong mather that don't benefit the hole js community.
To stop not only yourself from using `import('.')` you need to stop others as well so they are forced to write `import('./mod')` or better yet `import('./mod.js')`

**Comment by rluvaton** [comment]
> To stop not only yourself from using `import('.')` you need to stop others as well so they are forced to write `import('./mod')` or better yet `import('./mod.js')`
> 

What is the disadvantage of using `import(.)` when using the `exports` as @giltayar suggested?

On one hand, When creating a package, using the `import(./mod)` (or `.js`) you force the developers to know the entry point which is just an implementation detail.

On the other hand, for contributors, they need to look at the package.json in order to know what is the entry point when you use the `exports`

So I'm on the fence here 🤔

**Comment by goldbergyoni** [comment]
@giltayar @elektronik2k5 @jimmywarting 

Great comments, interesting discussion, seems to me like there are two orthogonal discussions hidden here:

- The mechanics - Whether to use ESM-compatible examples, naming ("index" or other name), etc. Seems like we are in-sync about this: ESM syntax is in order and the name "index" is too opinionated now that it has no special meaning

- The overarching encapsulation idea - Syntax aside, this bullet is mostly about exporting/grouping all the relevant objects/functions into a clear interface-like so the package consumer won't have to dig deep inside to find the "public" and useful functions she needs. Does this still make sense?

**Comment by jimmywarting** [comment]
I'm mostly just for "don't treat, index or extension less" imports in any special way. Solve it the ESM way always! With extension.

> The overarching encapsulation idea - Syntax aside, this bullet is mostly about exporting/grouping all the relevant objects/functions into a clear interface-like so the package consumer won't have to dig deep inside to find the "public" and useful functions she needs. Does this still make sense?

File names don't weight in as equally important for me at least. Use index if you feel like using it. it can have special meaning in terms of a web server where someone wants to serve a index page generated on the server side using some routing mechanism.

I would put it as a Documentation suggestion/best practices to not use index.js rather than throwing some eslint error for using the name index.js

**Comment by giltayar** [comment]
> p.s. Gil, I didn't get what specifically won't work with ESM, my example uses a direct file

Yup, you're right. It will work (in both CJS and ESM).

> The overarching encapsulation idea - Syntax aside, this bullet is mostly about exporting/grouping all the relevant objects/functions into a clear interface

You're right. This is the important thing. And it can be implemented in CJS and ESM using `exports` in `package.json`. So I would keep the practice, but change the mechanism to use `exports` in `package.json instead of a folder with `index.js` (which won't work in ESM) or using `../index.js` (which is ugly).

**Comment by goldbergyoni** [comment]
@giltayar Yes, this was exactly my intention, thanks

@jimmywarting You didn't explain why the encapsulation idea is, which is more strategic than the mechanics, doesn't appeal to you
